### PR TITLE
samples: drivers: espi: Convert espi sample to devicetree

### LINKED ISSUE #22344: convert espi sample to devicetree
samples/drivers/espi is an implementation of a driver for the MEC15xx eSPI peripheral/infrastructure.  PR #22154 includes a partial conversion of the GPIO capability of this infrastructure to devicetree.  Due to schedule and the amount of rework necessary the immediate problem was solved by #22298, and #22154 closed.

This PR exists to track the remaining task of converting this sample to use devicetree, or providing an actual driver implementation for this hardware, whichever seems appropriate.

### PR DESCRIPTION
**Pull request by albertofloyd**
Update sample driver test to use dts instead of kconfig.
Fixes #22344

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
A number of changes needed here, however we need to start with getting the DTS / yaml correct.  Trying to understand these GPIO pins in more detail.  For example, why are they not under the espi node?  and part of `microchip,xec-espi.yaml`?

**Comment by albertofloyd** [comment]
> A number of changes needed here, however we need to start with getting the DTS / yaml correct. Trying to understand these GPIO pins in more detail. For example, why are they not under the espi node? and part of `microchip,xec-espi.yaml`?

Fair enough, let me explain. 
These are no part of the SoC eSPI block/pins (eSPI bus only includes: RST, CLK, CS, IO0-3 which pin location is fixed within a Soc) 

These are board gpios used to interact with another system to monitor board power status in eSPI bus readiness context, these are not fixed in a board any GPIO can do.

**Comment by galak** [comment]
> Fair enough, let me explain.
> These are no part of the SoC eSPI block/pins (eSPI bus only includes: RST, CLK, CS, IO0-3 which pin location is fixed within a Soc)
> 
> These are board gpios used to interact with another system to monitor board power status in eSPI bus readiness context, these are not fixed in Soc and any GPIO can do.

which board power?  The board the SoC that the eSPI controller is on?  Or a board connected via eSPI?

bus readiness seems like a gray area between the controller and device(s) being talked to over it.

what is it that we are talking to over eSPI?  Are these GPIOs more a property of the device being connected to?  (I ask because I know various I2C sensors have similar concepts and utilize GPIO pins for that purpose).

**Comment by albertofloyd** [comment]
> > Fair enough, let me explain.
> > These are no part of the SoC eSPI block/pins (eSPI bus only includes: RST, CLK, CS, IO0-3 which pin location is fixed within a Soc)
> > These are board gpios used to interact with another system to monitor board power status in eSPI bus readiness context, these are not fixed in Soc and any GPIO can do.
> 
> which board power? The board the SoC that the eSPI controller is on? Or a board connected via eSPI?

As you know, boards have more than one power rails, not all of them used to power the SoC.
SoC also have different different power rails used to power the different HW blocks.
These pins allow these specific-boards acting as eSPI slaves to know when eSPI master has a specific board power rail stable which is not exactly the powering the eSPI HW block.


> bus readiness seems like a gray area between the controller and device(s) being talked to over it.

This is not bus readiness, it's board power rails stability and overall SoC readiness. 
 
> what is it that we are talking to over eSPI? 
eSPI bus is a very flexible bus where you can tunnel all kind of traffic: (Virtual GPIOs, VIrtual I2C, PS2, Keyboard, MIDI, etc) 

> Are these GPIOs more a property of the device being connected to? (I ask because I know various I2C sensors have similar concepts and utilize GPIO pins for that purpose).

Yes, these GPIOS are device properties and not related to the bus. THese are more like LEDs/buttons than bus signals. That's the main reasons I didn't put these as part eSPI HW controller device tree.
Some additional ones to consider

1) Board power rails are not common across all boards or SoC that implement eSPI HW controller.
2) These pins are optional, and considered **generic board pins** more similar to buttons or LEDS and do not belong in the eSPI HW block.

Note that original intention of this eSPI sample code is to exemplifies how to use thes pins in MEC15xx boards before the actual eSPI driver is tested.

Let me know if the naming convention is causing this confusion and/or if this approach to handle board specific features like this one is not generic enough.

**Comment by galak** [comment]
So it seems these should be more board defined GPIOs and having a compatible related to that would make more sense.

So I think something like:

```
    board-power {
        compatible = "microchip,mec15xx-board-power";
           pwrg-gpios = <&gpioa 10 0>;
           rsm-gpios = <&gpiob 12 0>;
      };
```

Have the yaml in dts/bindings/boards as you want to share this between several boards.   The other case / example will be having the binding in a board specific dir. (<BOARD_DIR>/dts/bindings/board.yaml).

**Comment by albertofloyd** [comment]
> pwrg

For given option 1), do you mean to create a file in here ZEPHYR_PATH>/dts/bindings/boards/mec15xx-board-power.yaml?

**Comment by galak** [comment]
> For given option 1), do you mean to create a file in here ZEPHYR_PATH>/dts/bindings/boards/mec15xx-board-power.yaml?

Yes, since I expect you'll use the binding on more than one board.

> Also isn't this compatible string "mec15xx-board-power " a bit long? Compiling this results in
> DT_MICROCHIP_MEC15XX_BOARD_POWER_PINS_PWRG_GPIOS_CONTROLLER

Feel free to suggest another name, I was just 

Also, with the new macro's this would be `DT_GPIO_LABEL(DT_INST(0, microchip_mec15xx_board_power))`

**Comment by albertofloyd** [comment]
@galak I have added move the dts bindings definition file and perform the modifications in both boards.  However, looking into DT macro documentation and into current samples (blinky/usb_stm32 driver).  I have been unable to figure how to access the node that is not an alias neither part of a SoC hw block. 

Take a look in the updated to see what I may be misunderstanding in the usage.

File: boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts [hunk 1]
```diff
@@ -19,10 +19,19 @@
...
+
+		/* MCHP_GPIO_0054 */
+		rsm-gpios = <&gpiob 12 GPIO_ACTIVE_HIGH>;
+	};
+
 	aliases {
 		pwm-0 = &pwm0;
 		peci-0 = &peci0;
-
+		espi-0 = &espi0;
```
Comment (galak) [inline boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts]: Is this change needed?

File: samples/drivers/espi/sample.yaml [hunk 2]
```diff
@@ -1,8 +1,13 @@
 sample:
+  description: A basic demo to showcase eSPI driver usage and handshake with
+    eSPI master
   name: eSPI driver sample
 tests:
   sample.drivers.espi:
-    tags: drivers
+    tags: drivers espi
+    filter: dt_compat_enabled_with_alias("gpio-espi", "gpioespi0") and
```
Comment (galak) [inline samples/drivers/espi/sample.yaml]: filter here should just be something like `dt_compat_enabled("microchip,mec15xx-board-power")`

File: samples/drivers/espi/src/main.c [hunk 3]
```diff
@@ -47,10 +47,35 @@ struct oob_response {
...
 
-#ifdef CONFIG_ESPI_GPIO_DEV_NEEDED
-static struct device *gpio_dev0;
-static struct device *gpio_dev1;
 #define PWR_SEQ_TIMEOUT    3000u
+
+/* The devicetree node identifier for the board power rails pins. */
+#define BRD_PWR_NODE DT_INST(0, microchip_mec15xx_board_power)
+
+#ifdef USING_DT_MACROS
```
Comment (galak) [inline samples/drivers/espi/src/main.c]: When is this not true?

**Review by galak — DISMISSED** [review]
In general this looks good.  

Do you see the sample running on some board beyond the 2 that it currently supports?

File: samples/drivers/espi/src/main.c (hunk 3 above)
Comment (albertofloyd) [reply samples/drivers/espi/src/main.c]: This should always be true and nothing should be done in the else.
However when I use the new DT macros, the sample doesn't compile.
Could you take a look what could be wrong in the node access?

File: boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts (hunk 1 above)
Comment (albertofloyd) [reply boards/arm/mec1501modular_assy6885/mec1501modular_assy6885.dts]: Removed

**Comment by albertofloyd** [comment]
> In general this looks good.
> 
> Do you see the sample running on some board beyond the 2 that it currently supports?

No other board supports eSPI other than MEC15xx boards

**Comment by albertofloyd** [comment]
@galak re-review.
I have changed the approach and use an overlay in sample itself to get it working with new DT macros.

**Comment by albertofloyd** [comment]
ping @galak

**Comment by albertofloyd** [comment]
ping @nashif , @galak

**Review by nashif — APPROVED** [review]

**Comment by albertofloyd** [comment]
ping @galak

**Review by dcpleung — APPROVED** [review]
